WIP: Blog post for changed block tracking feature (#48456)
kubernetes/website: content/en/blog/_posts/2024-xx-xx-csi-changed-block-tracking.md

---
layout: blog

> title: 'Kubernetes v1.33: Changed Block Tracking (CBT )API support (alpha)'

Reviewer: `suggestion
title: 'Kubernetes v1.33: Changed Block Tracking API support (alpha)'
```

> date: 2025-XX-XX

Reviewer 2: Although the schedule is going to be a bit tight, we would like this to get released on Tuesday, 13th May. The timestamp helps schedule the publication timing to be consistent with other blogs.

```suggestion
date: 2025-05-13T10:30:00-08:00
```

draft: true

> slug: csi-changed-block-tracking

Reviewer 2: For feature blogs, we would like to keep the same prefix for consistency.

```suggestion
slug: kubernetes-v1-33-csi-changed-block-tracking
```

Reviewer: Please update to match the suggestion.

author: >
   Prasad Ghangal (Veeam Kasten)
---

> We're excited to announce that Kubernetes v1.33 introduces alpha support for the **Changed Block Tracking (CBT) API**. This new feature enhances the Kubernetes storage ecosystem by providing an efficient way to identify changed blocks in block volume snapshots, enabling faster and more resource-efficient backup operations.

Reviewer: Watch out for ambiguous "we"

I would use wording such as:
```suggestion
I'm excited to announce that Kubernetes v1.33 introduced alpha support for a _changed block tracking_ mechanism.
This enhances the Kubernetes storage ecosystem by providing an efficient way for storage drivers to identify changed blocks in PersistentVolume snapshots.
With a driver that can use the feature, you could benefit from faster and more resource-efficient backup operations.
```

Author: @lmktfy we have added multiple authors, so I think we is correct?

## What is changed block tracking?

> The Changed Block Tracking API introduces a new Container Storage Interface (CSI) capability that allows you to:

Reviewer: ```suggestion
The improvement is a change to the Container Storage Interface (CSI), and also to the storage snapshot support in Kubernetes itself.
With the alpha features enabled, your cluster can:
```

- Identify allocated blocks within a CSI volume snapshot
- Determine changed blocks between two snapshots of the same volume

> For Kubernetes users managing large datasets, this means significantly more efficient backup processes, as backup applications can now focus only on blocks that have actually changed rather than processing entire volumes.

Reviewer 2: We could perhaps use some slight rewording for better clarity?

```suggestion
For Kubernetes users managing large datasets, this API enables significantly more efficient backup processes. Backup applications can now focus only on the blocks that have changed, rather than processing entire volumes.
```

Reviewer: I am wary of highlighting that it's an API. CSI is the interface (and PersistentVolume, VolumeSnapshot) etc. CBT is not really an API in its own right.

> >  NOTE:

Reviewer: Don't use a block quote when not quoting.

>  As of now, the Changed Block Tracking API is supported only for block volumes and not for file volumes. CSI drivers that manage file-based storage systems will not be able to implement this capability.

> ## Why add changed block tracking API to Kubernetes

Reviewer: Ideally, reword this.

As Kubernetes adoption grows for stateful workloads managing critical data, the need for efficient backup solutions becomes increasingly important. Traditional full backup approaches face challenges with:

- Long backup windows
- High resource utilization

> - Increased storage costs

Reviewer 2: I think it would be useful to provide more details for each point, perhaps with some example scenarios where this becomes a challenge.

> The CBT API addresses these challenges by providing native Kubernetes support for incremental backup capabilities through the CSI interface.

Reviewer 2: It seems the "CBT" acronym is used only once here. Although the acronym is already clarified earlier in the blog, it may be worth writing out "Changed Block Tracking" to make it consistent with the rest of the article.

Reviewer: Again, I am wary of framing it as an API.

## Key components

The implementation consists of three primary components:

1. **CSI SnapshotMetadata Service API**: A gRPC service that provides information about volume snapshots, including changed block data
2. **SnapshotMetadataService Custom Resource**: Advertises the availability of a CSI driver's metadata service along with connection details

> 3. **ExternalSnapshotMetadata Sidecar**: Implements the server-side of the API, bridging CSI drivers with backup applications

Reviewer: Odd capitalization. I think I would use sentence case italics.

> ## Vendor implementation requirements

Reviewer: Storage drivers are not always written in exchange for money.

### Storage vendor responsibilities

Storage vendors who want to support the changed block tracking feature must implement specific requirements:

1. **Implement CSI RPCs**: Storage vendors need to implement the `SnapshotMetadata` service as defined in the [CSI specifications protobuf](https://github.com/container-storage-interface/spec/blob/master/csi.proto). This service requires server-side streaming implementations for the following RPCs:

   - `GetMetadataAllocated`: For identifying allocated blocks in a snapshot
   - `GetMetadataDelta`: For determining changed blocks between two snapshots

2. **Storage backend capabilities**: Ensure the storage backend has the capability to track and report block-level changes.

3. **Deploy external components**: Integrate with the `external-snapshot-metadata` sidecar to expose the snapshot metadata service.

4. **Register custom resource**: Create and maintain a `SnapshotMetadataService` custom resource that advertises the availability of their metadata service and provides connection details.

5. **Support error handling**: Implement proper error handling for these RPCs according to the CSI specification requirements.

> ### Backup vendor responsibilities

Reviewer: Softwaee need not be written for money; vendor implies money.

Backup vendors looking to leverage this feature must:

> 1. **Setup authentication**: Configure proper authentication for clients accessing the storage system using Kubernetes ServiceAccount tokens.

Reviewer 2: To follow the above sentence, this should be in the verb form

```suggestion
1. **Set up authentication**: Configure proper authentication for clients accessing the storage system using Kubernetes ServiceAccount tokens.
```

3. _Deploy external components_: Integrate with the `external-snapshot-metadata` sidecar to expose the snapshot metadata service.

4. _Register custom resource_: Create and maintain a `SnapshotMetadataService` custom resource that advertises the availability of their metadata service and provides connection details.

5. _Support error handling_: Implement proper error handling for these RPCs according to the CSI specification requirements.

### Backup client responsibilities

Backup vendors looking to leverage this feature must:

1. **Set up authentication**: Configure proper authentication for clients accessing the storage system using Kubernetes ServiceAccount tokens.

2. **Implement streaming client-side code**: Develop clients that implement the streaming gRPC APIs defined in the [schema.proto](https://github.com/kubernetes-csi/external-snapshot-metadata/blob/main/proto/schema.proto) file. Specifically:
   - Implement streaming client code for `GetMetadataAllocated` and `GetMetadataDelta` methods
   - Handle server-side streaming responses efficiently as the metadata comes in chunks

>    - Process the `SnapshotMetadataResponse` message format with proper error handling

Reviewer 3: I suggest adding that the external-snapshot-metadata repository provides a convenient [iterator](https://github.com/kubernetes-csi/external-snapshot-metadata/tree/master/pkg/iterator) support package that handles all of the above.

---
layout: blog
title: 'Announcing Changed Block Tracking API support (alpha)'
date: 2025-05-14T18:30:00+05:30
draft: true
slug: csi-changed-block-tracking
author: >
   [Prasad Ghangal](https://github.com/PrasadG193) (Veeam Kasten)
   [Carl Braganza](https://github.com/carlbraganza) (Veeam Kasten)
---

> We're excited to announce the alpha support for a _changed block tracking_ mechanism. This enhances the Kubernetes storage ecosystem by providing an efficient way for storage drivers to identify changed blocks in PersistentVolume snapshots. With a driver that can use the feature, you could benefit from faster and more resource-efficient backup operations.

Reviewer: - Split this over more than one line in the Markdown source
- we should mention "CSI" here, and that mention should be a hyperlink to https://kubernetes.io/docs/concepts/storage/volumes/#csi (which then links to docs for the CSI subproject)

> 

Reviewer: Let's help [readers who are keen] skip to the Getting Started heading.

{{< /note >}}

## Benefits of changed block tracking support in Kubernetes

As Kubernetes adoption grows for stateful workloads managing critical data, the need for efficient backup solutions becomes increasingly important. Traditional full backup approaches face challenges with:

- _Long backup windows_: Full volume backups can take hours for large datasets, making it difficult to complete within maintenance windows.
- _High resource utilization_: Backup operations consume substantial network bandwidth and I/O resources, especially for large data volumes and data-intensive applications.
- _Increased storage costs_: Repetitive full backups store redundant data, causing storage requirements to grow linearly even when only a small percentage of data actually changes between backups.

The Changed Block Tracking API addresses these challenges by providing native Kubernetes support for incremental backup capabilities through the CSI interface.

## Key components

The implementation consists of three primary components:

> 1. _CSI SnapshotMetadata Service API_: A gRPC service that provides information about volume snapshots, including changed block data

Reviewer: ```suggestion
1. _CSI SnapshotMetadata service_: An API, offered via gRPC, that provides information about volume snapshots, including changed block data.
```
?

> 2. _SnapshotMetadataService Custom Resource_: Advertises the availability of a CSI driver's metadata service along with connection details

Reviewer: ```suggestion
1. _SnapshotMetadataService API_: If you install this API using its CustomResourceDefinition,
   the CSI snapshot metadata controller advertises the availability of
   a CSO driver's metadata service, along with connection details.
```
Use auto numbering.

> 3. _External Snapshot Metadata Sidecar_: Implements the server-side of the API, bridging CSI drivers with backup applications

Reviewer: Does it have to be a sidecar?

Reviewer 4: Yes, it has to be a sidecar to CSI driver pods

## Implementation requirements

### Storage provider responsibilities

> Storage providers who want to support the changed block tracking feature must implement specific requirements:

Reviewer: ```suggestion
If you're an author of a storage integration with Kubernetes, and you
want to support the changed block tracking feature, you must implement specific requirements:
```

1. _Implement CSI RPCs_: Storage providers need to implement the `SnapshotMetadata` service as defined in the [CSI specifications protobuf](https://github.com/container-storage-interface/spec/blob/master/csi.proto). This service requires server-side streaming implementations for the following RPCs:

   - `GetMetadataAllocated`: For identifying allocated blocks in a snapshot
   - `GetMetadataDelta`: For determining changed blocks between two snapshots

2. _Storage backend capabilities_: Ensure the storage backend has the capability to track and report block-level changes.

3. _Deploy external components_: Integrate with the `external-snapshot-metadata` sidecar to expose the snapshot metadata service.

> 4. _Register custom resource_: Create and maintain a `SnapshotMetadataService` custom resource that advertises the availability of their metadata service and provides connection details.

Reviewer: You mean CustomResourceDefinition, here.

5. _Support error handling_: Implement proper error handling for these RPCs according to the CSI specification requirements.

### Backup client responsibilities

> Backup clients looking to leverage this feature must:

Reviewer: "client" isn't quite right here. Is the thing in question a client of:

- Kubernetes
- a backup server
- something else? 

Try "solution" or "tool" rather than "client".

1. _Set up authentication_: The backup application must provide a Kubernetes ServiceAccount token when using the
[Kubernetes SnapshotMetadataService API](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3314-csi-changed-block-tracking#the-kubernetes-snapshotmetadata-service-api).

> Appropriate RBAC permission must be established to authorize the backup application ServiceAccount to obtain such tokens.

Reviewer: ```suggestion
Appropriate access grants, such as RBAC RoleBindings, must be established to authorize the backup application ServiceAccount to obtain such tokens.
```
- We can infer they use a ServiceAccount (but they might not) 
- We can't assume that every cluster uses K8s RBAC

   - Implement streaming client code for `GetMetadataAllocated` and `GetMetadataDelta` methods
   - Handle server-side streaming responses efficiently as the metadata comes in chunks
   - Process the `SnapshotMetadataResponse` message format with proper error handling

   The `external-snapshot-metadata` GitHub repository provides a convenient [iterator](https://github.com/kubernetes-csi/external-snapshot-metadata/tree/master/pkg/iterator) support package that can be used to implement the client.

3. _Handle large dataset streaming_: Design clients to efficiently handle large streams of block metadata that could be returned for volumes with significant changes.

4. _Optimize backup processes_: Modify backup workflows to use the changed block metadata to identify and only transfer changed blocks to make backups more efficient, reducing both backup duration and resource consumption.

## Getting started

To use changed block tracking in your cluster:

1. Ensure your CSI driver supports volume snapshots and implements the snapshot metadata capabilities with the required `external-snapshot-metadata` sidecar

> 2. Make sure the `SnapshotMetadataService` custom resource is registered using CRD

Reviewer: No backticks around the names of API kinds.

---
layout: blog
title: 'Announcing Changed Block Tracking API in Kubernetes v1.33'

> date: 2024-XX-XX

Reviewer: Try this (you need to mark the article as draft)
```
date: 2025-XX-XX
draft: true
```

slug: csi-changed-block-tracking

> ---

Reviewer: Make sure to list an article author.

We're excited to announce that Kubernetes v1.33 introduces alpha support for the **Changed Block Tracking (CBT) API**. This new feature enhances the Kubernetes storage ecosystem by providing an efficient way to identify changed blocks in volume snapshots, enabling faster and more resource-efficient backup operations.

> ## What is Changed Block Tracking?

Reviewer: Headings in sentence case, if you can.

The Changed Block Tracking API introduces a new Container Storage Interface (CSI) capability that allows you to:

- Identify allocated blocks within a CSI volume snapshot
- Determine changed blocks between two snapshots of the same volume

For Kubernetes users managing large datasets, this means significantly more efficient backup processes, as backup applications can now focus only on blocks that have actually changed rather than processing entire volumes.

> ## Why We Built This

Reviewer: I recommend rewording - see https://kubernetes.io/docs/contribute/blog/article-submission/#article-content about the word “we”

As Kubernetes adoption grows for stateful workloads managing critical data, the need for efficient backup solutions becomes increasingly important. Traditional full backup approaches face challenges with:

- Long backup windows
- High resource utilization
- Increased storage costs

The CBT API addresses these challenges by providing native Kubernetes support for incremental backup capabilities through the CSI interface.

> ## Key Components

Reviewer: Headings in sentence case, if you can.

The implementation consists of three primary components:

1. **CSI SnapshotMetadata Service API**: A gRPC service that provides information about volume snapshots, including changed block data

> 2. **SnapshotMetadataService Custom Resource**: Advertises the availability of a CSI driver's metadata service along with connection details

Reviewer: Is this CRD documented? We should avoid promoting an undocumented API.

Author: @lmktfy we have published CRD here https://github.com/kubernetes-csi/external-snapshot-metadata/blob/main/client/config/crd/cbt.storage.k8s.io_snapshotmetadataservices.yaml
Are you suggesting to include that in the blog? We also have published the documentation at https://kubernetes-csi.github.io/docs/external-snapshot-metadata.html

Reviewer: OK, link to the CSI docs page for SnapshotMetadataService.

3. **ExternalSnapshotMetadata Sidecar**: Implements the server-side of the API, bridging CSI drivers with backup applications

> ## Getting Started

Reviewer: Headings in sentence case, if you can.

To use the Changed Block Tracking API in your cluster:

1. Ensure your CSI driver supports volume snapshots and implements the snapshot metadata capabilities with the required `external-snapshot-metadata` sidecar
2. Verify the presence of a SnapshotMetadataService Custom Resource for your CSI driver
3. Create clients that can access the API using appropriate authentication (via Kubernetes ServiceAccount tokens)

The API provides two main functions:

- `GetMetadataAllocated`: Lists blocks allocated in a single snapshot
- `GetMetadataDelta`: Lists blocks changed between two snapshots

> ## What’s next?

Reviewer: Headings in sentence case, if you can.

> Depending on feedback and adoption, the Kubernetes team plans to push the CSI Snapshot Metadata implementation to Beta in the future releases.

Reviewer: ```suggestion
Depending on feedback and adoption, the Kubernetes developers hope to push the CSI Snapshot Metadata implementation to Beta in the future releases.
```
Avoid implying a promise here.

> ## Where can I learn more?

Reviewer: Headings in sentence case, if you can.

For those interested in trying out this new feature:

> - The [design proposal](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3314-csi-changed-block-tracking) for the snapshot metadata feature.

Reviewer: It's an enhancement proposal, not a design proposal.

- Complete gRPC protocol definitions for snapshot metadata API: [schema.proto](https://github.com/kubernetes-csi/external-snapshot-metadata/blob/main/proto/schema.proto)
- Example snapshot metadata client implementation: [snapshot-metadata-lister](https://github.com/kubernetes-csi/external-snapshot-metadata/tree/main/examples/snapshot-metadata-lister)
- End-to-end example with csi-hostpath-driver: [example documentation](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/docs/example-ephemeral.md)

## How do I get involved?

> This project, like all of Kubernetes, is the result of hard work by many contributors from diverse backgrounds working together. On behalf of SIG Storage, I would like to offer a huge thank you to the contributors who helped review the design and implementation of the project, including but not limited to the following:

Reviewer: ```This enhancement, like all of Kubernetes, is the result of hard work by many contributors from diverse backgrounds working together. On behalf of SIG Storage, I would like to offer a huge thank you to the contributors who helped review the design and implementation of the project, including but not limited to the following:
```

Reviewer: Hmm, I don't know!

- Ben Swartzlander ([bswartz](https://github.com/bswartz))
- Carl Braganza ([carlbraganza](https://github.com/carlbraganza))
- Daniil Fedotov ([hairyhum](https://github.com/hairyhum))
- Ivan Sim ([ihcsim](https://github.com/ihcsim))
- Nikhil Ladha ([Nikhil-Ladha](https://github.com/Nikhil-Ladha))
- Prasad Ghangal ([PrasadG193](https://github.com/PrasadG193))
- Praveen M ([iPraveenParihar](https://github.com/iPraveenParihar))
- Rakshith R ([Rakshith-R](https://github.com/Rakshith-R))
- Xing Yang ([xing-yang](https://github.com/xing-yang))

We also want to thank everyone else who has contributed to the project, including others who helped review the [KEP](https://github.com/kubernetes/enhancements/pull/4082) and the [CSI spec PR](https://github.com/container-storage-interface/spec/pull/551)

For those interested in getting involved with the design and development of CSI or any part of the Kubernetes Storage system, join the [Kubernetes Storage Special Interest Group](https://github.com/kubernetes/community/tree/master/sig-storage) (SIG). We always welcome new contributors.

> We also hold regular [Data Protection Working Group meetings](https://docs.google.com/document/d/15tLCV3csvjHbKb16DVk-mfUmFry_Rlwo-2uG6KNGsfw/edit). New attendees are welcome to join our discussions.

Reviewer: I recommend rewording - see https://kubernetes.io/docs/contribute/blog/article-submission/#article-content about the word “we”

---
layout: blog

> title: 'Announcing Changed Block Tracking API in Kubernetes v1.33'

Reviewer: Usually we follow a specific format for the article title.

Reviewer 5: Yep, @PrasadG193 I suggest this option to align with other feature blog released for 1.33:

```suggestion
title: 'Kubernetes v1.33: Changed Block Tracking (CBT )API support (alpha)'
```

date: 2025-XX-XX
draft: true
slug: csi-changed-block-tracking
author: >
   Prasad Ghangal (Veeam Kasten)
---

We're excited to announce that Kubernetes v1.33 introduces alpha support for the **Changed Block Tracking (CBT) API**. This new feature enhances the Kubernetes storage ecosystem by providing an efficient way to identify changed blocks in volume snapshots, enabling faster and more resource-efficient backup operations.

## What is changed block tracking?

The Changed Block Tracking API introduces a new Container Storage Interface (CSI) capability that allows you to:

- Identify allocated blocks within a CSI volume snapshot
- Determine changed blocks between two snapshots of the same volume

For Kubernetes users managing large datasets, this means significantly more efficient backup processes, as backup applications can now focus only on blocks that have actually changed rather than processing entire volumes.

> 

Author: I have updated introduction also added a note about it

## Why add changed block tracking API to Kubernetes

As Kubernetes adoption grows for stateful workloads managing critical data, the need for efficient backup solutions becomes increasingly important. Traditional full backup approaches face challenges with:

- Long backup windows
- High resource utilization
- Increased storage costs

The CBT API addresses these challenges by providing native Kubernetes support for incremental backup capabilities through the CSI interface.

## Key components

The implementation consists of three primary components:

1. **CSI SnapshotMetadata Service API**: A gRPC service that provides information about volume snapshots, including changed block data
2. **SnapshotMetadataService Custom Resource**: Advertises the availability of a CSI driver's metadata service along with connection details
3. **ExternalSnapshotMetadata Sidecar**: Implements the server-side of the API, bridging CSI drivers with backup applications

## Getting started

> To use the Changed Block Tracking API in your cluster:

Reviewer: ```suggestion
To use changed block tracking in your cluster:
```

1. Ensure your CSI driver supports volume snapshots and implements the snapshot metadata capabilities with the required `external-snapshot-metadata` sidecar

> 2. Verify the presence of a SnapshotMetadataService Custom Resource for your CSI driver

Reviewer: Do you need to deploy a particular CRD first?
Markdown counts for you, so you could try: 
```suggestion
1. Verify the presence of a SnapshotMetadataService custom resource for your CSI driver
```

3. Create clients that can access the API using appropriate authentication (via Kubernetes ServiceAccount tokens)

The API provides two main functions:

- `GetMetadataAllocated`: Lists blocks allocated in a single snapshot
- `GetMetadataDelta`: Lists blocks changed between two snapshots

> 

Reviewer 6: Can you add what a storage vendor needs to do and what a backup vendor needs to do to support this feature?

Reviewer 6: Is this comment addressed anywhere?

Author: I have added implementation requirements
